### DISCUSSION TITLE
Protected static pages possible?

### DISCUSSION THREAD

**Discussion by jozsefm** [vercel/next.js#20185] [Help]
### Problem
Im trying to solve a seemingly simple problem. I want to allow only logged in users to access a bunch of static pages of a Next.js app. [_edit: for clarification this is not the whole site, just some select routes_] To my surprise all the approaches I tried seems to fall short. I will describe them here, but I didn't find any way to achieve this so if you know how to do it reading my struggles won't give more context.

### Story
If I create an SSG page which is conditionally rendering it's content, I'm not fully confident what happens behind the scenes. What I saw [here](https://www.reddit.com/r/nextjs/comments/jmqlys/comment/gawxzww) seems to suggest the pregenrated code includes the content of both the true and false cases which is not acceptable for me. Even if the displayed information is showing what an unauthorized person should see, if the paid content is delivered in the source code it will be crawled and scrapped or a more savvy user will simply take it. But even if only the non logged in case is delivered, there's no much use of pregenerating that loading screen..

So the real solution I thought will be to use the vercel config to set up a redirect that maps the generic route of the content to an API function that will check the session from a cookie (I'm using next-iron-session) and let the request pass to the route only if the user is logged in. Well it sounds good, but doesn't work this way. There's no such thing as "letting the request pass through". I have to return whatever content I want right there, and I dont know how can I grab the specific page from there as a file/bytestream/anything and send it back. I don't think it's possible nor a good solution even if doable. I could redirect the request and I pondered with many tricks here, like redirecting to a very secret url, match that with a regex in the config and map that back to a normal url there, but I couldn't get it working. And again, that would be a major hack. 

### Conclusion
Now I think ideally we should be able to intercept any request to a given route with a serverless function and decide to let it go through or reject or manipulate it however we want. I know an SSR page could provide the functionality I'm looking for, but I don't want to re-render the same static content for 1000s of users continuously because of an authentication check. Is it a currently achievable idea or if not, would anybody else find such a feature useful?

**Comment by Vadorequest** [comment]
Maybe this might help:
https://vercel.com/blog/protecting-deployments

It's not free, but it might help nonetheless.

See https://github.com/vercel/next.js/discussions/12130

**Reply by jozsefm** [reply]
Thanks @Vadorequest for the suggestion, but my use case is different. I'm creating a website selling an online course. It has public marketing pages and protected (behind authentication) parts for the course content. I can't use a whole deployment for restricting access, it has to live under the same domain name. I'm a bit surprised that it's not a well supported use case. I honestly wonder if I'm trying to solve the wrong problem here, or else why didn't this came up already...

**Reply by Vadorequest** [reply]
I did some research on this yesterday, and found those:
https://github.com/flawyte/vercel-basic-auth
**https://github.com/flawyte/static-auth**

But, I wasn't able to actually use it. Documentation is rather bad for the Vercel integration, maybe it only works for static sites. But, it might be exactly what you need @jozsefm @andrzejwalk 

https://github.com/flawyte/static-auth/issues/3

**Reply by jozsefm** [reply]
Thanks again for the effort! I didn't see these previously but they seem to use Vercel simply as a host for static resources. I'm not sure if I got that right, but I don't see how those patterns could be implemented for a Next.js app. API functions definitely doesn't work this way and I want to avoid the trouble of writing a custom server for this at all costs.

**Reply by Vadorequest** [reply]
Yeah, which is the case if you use `next export`, you get such a static export and can use flawyte/static-auth in such case.

But if you're doing it like me, like using the Next.js `serverless` mode, then I haven't found a way either. I thought that'd be simple...

Unrelated, but my next candidate is https://magic.link/ to achieve what I want, for free. (internal use, less than 100 users)

**Reply by jozsefm** [reply]
True, in case of an exported app it might do the trick. I would really like to have incremental regeneration enabled for the static pages as I expect frequent changes for them (daily so no real need for SSR again) and in addition I want to deploy a few API functions, so I think I can't really have it all now. I thought it would be easy too...

My next candidate would be to use something like: https://fab.dev/guides/adding-server-side-logic#checking-a-cookie-before-allowing-a-request-to-proceed or https://developers.cloudflare.com/workers/examples/auth-with-headers but I don't really want to give up on all the benefits of Vercel and Next.js

**Comment by andrzejwalk** [comment]
I'm also interested in finding a solution for exactly the same scenario: protected static pages.

**Comment by mattvick** [comment]
Hello, have you guys considered deploying your SSG page to S3 and protecting these using Basic HTTP Authentication?

[This article](https://medium.com/hackernoon/serverless-password-protecting-a-static-website-in-an-aws-s3-bucket-bfaaa01b8666) explains how to protect S3 assets using CloudFront and Lambda@Edge, and YouTube has a [video tutorial](https://www.youtube.com/watch?v=gc3w_bMtcQE).

> Lambda@Edge allows you to attach AWS Lambda functions to CloudFront behaviors ... Lambda@Edge now [officially supports](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/lambda-requirements-limits.html#lambda-requirements-network-access) network access for viewer request/response events. It is now totally possible to manage credentials in a DynamoDB table or use any external API/provider to authenticate users, as long as the network call takes [less than 5s to complete](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/cloudfront-limits.html#limits-lambda-at-edge).

@jozsefm regarding some specific points you made:

> "just some select routes" ... "it has to live under the same domain name"

CloudFront can direct select routes to S3 and other routes under the same domain name to Vercel.

> " I would really like to have incremental regeneration enabled for the static pages as I expect frequent changes for them (daily so no real need for SSR again)"

I imagine it would be possible to schedule SSG regeneration and deployment to S3 to happen every 24 hours using [cron jobs on Vercel](https://vercel.com/guides/how-to-setup-cron-jobs-on-vercel).
